src/components/Editor.jsx

```diff
@@ -142,11 +148,50 @@ class Editor extends React.Component {
     this._resizeEditor();
   }
 
+  _handleKeyPress(event) {
+    if (event.key === 'i' && (event.metaKey || event.ctrlKey) &&
+        !event.altKey && !event.ctrlKey) {
+      const {session, selection} = this._editor;
+      const doc = session.getDocument();
+      const range = selection.getRange();
+
+      const options = {};
+      if (session.getUseSoftTabs()) {
+        options.indent_char = ' ';
+        options.indent_size = session.getTabSize();
+      } else {
+        options.indent_char = '\t';
+        options.indent_size = 1;
+      }
+
+      const startIndex = doc.positionToIndex(range.start);
+      const endIndex = doc.positionToIndex(range.end);
+
+      const value = session.getValue();
+
+      const [newValue, newStartIndex, newEndIndex] =
+        format(value, startIndex, endIndex, this.props.language, options);
+
+      // Make sure we set the new value for the editor via editor.setValue
+      // instead of session.setValue otherwise the undo/redo stack will be
+      // clobbered!
```

Comment:
Hmmm self documenting code is pretty awesome. But, I would still want to have this comment inside of `_updateEditorValueWithoutClobberingUndoStack`.

eg:

```
_updateEditorValueWithoutClobberingUndoStack(value) {
  // Make sure we set the new value for the editor via editor.setValue
  // instead of session.setValue otherwise the undo/redo stack will be
  // clobbered!
  this._editor.setValue(value);
}
```

If I could change the API of ACE itself, then I agree it would be nice to have: `this._editor.setValue(newValue, {ignoreUndo: true})` or something similar

Reply:
I’m OK with this compromise—it’s a little jarring to see inline comments in the Popcode source since they are so rare. Also once it’s extracted into its own method, I would propose making it a method-level comment, since those are a little more firmly “attached” and thus a bit less risk of getting out of sync with the actual code.